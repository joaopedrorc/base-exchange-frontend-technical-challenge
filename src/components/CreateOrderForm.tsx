'use client';

import { z } from 'zod';
import { zodResolver } from '@hookform/resolvers/zod';
import { useForm } from 'react-hook-form';
import {
  Form,
  FormControl,
  FormDescription,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from '@/components/ui/form';
import {
  Select,
  SelectContent,
  SelectGroup,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from '@/components/ui/select';
import { Input } from '@/components/ui/input';
import { Button } from '@/components/ui/button';
import { useState } from 'react';

const formSchema = z.object({
  asset: z.string().min(1, 'Ativo é obrigatório'),
  side: z.enum(['Compra', 'Venda'], {
    required_error: 'Selecione uma opção',
  }),
  price: z
    .string()
    .regex(
      /^(\s?R\$)?\s?\d{1,3}(\.\d{3})*(,\d{2})?$/,
      'Preço inválido, use o formato R$ 1.000,00'
    )
    .min(1, 'Preço é obrigatório'),
  amount: z
    .string()
    .min(1, { message: 'Mínimo de 1 unidade' })
    .max(1000, { message: 'Máximo de 1000 unidades' }),
});

export function CreateOrderForm() {
  const [loading, setLoading] = useState(false);
  const form = useForm<z.infer<typeof formSchema>>({
    resolver: zodResolver(formSchema),
    defaultValues: {
      asset: '',
      side: 'Compra',
      price: '',
      amount: '',
    },
  });

  async function onSubmit(values: z.infer<typeof formSchema>) {
    try {
      setLoading(true);
      const res = await fetch('/api/orders', {
        method: 'POST',
        body: JSON.stringify(values),
      });

      const data = await res.json();
      console.log(data);
      setLoading(false);
    } catch (error) {
      console.log('ERROR', error);

      setLoading(false);
    }

    console.log(values);
  }

  function formatToBRL(value: string) {
    const numericValue = value.replace(/\D/g, '');
    const amount = parseInt(numericValue || '0', 10) / 100;
    return amount.toLocaleString('pt-BR', {
      style: 'currency',
      currency: 'BRL',
    });
  }

  function formatInteger(value: string) {
    let numericValue = value.replace(/\D/g, '');
    let intValue = parseInt(numericValue || '0', 10);

    if (intValue > 1000) {
      intValue = 1000;
    }

    return intValue > 0 ? intValue.toString() : '';
  }

  return (
    <div>
      <Form {...form}>
        <form
          onSubmit={form.handleSubmit(onSubmit)}
          className="grid grid-cols-1 gap-4 space-y-8 lg:grid-cols-4"
        >
          <FormField
            control={form.control}
            name="asset"
            render={({ field }) => (
              <FormItem>
                <FormLabel>Ações</FormLabel>
                <FormControl>
                  <Select
                    onValueChange={field.onChange}
                    defaultValue={field.value}
                  >
                    <SelectTrigger className="w-full">
                      <SelectValue placeholder="Selecione uma ação" />
                    </SelectTrigger>
                    <SelectContent>
                      <SelectGroup>
                        <SelectItem value="B3SA3">B3 (B3SA3)</SelectItem>
                        <SelectItem value="BBAS3">
                          Banco do Brasil (BBAS3)
                        </SelectItem>
                        <SelectItem value="PETR4">Petrobras (PETR4)</SelectItem>
                        <SelectItem value="HAPV3">Hapvida (HAPV3)</SelectItem>
                        <SelectItem value="ITUB4">
                          Itaú Unibanco (ITUB4)
                        </SelectItem>
                        <SelectItem value="BBDC4">Bradesco (BBDC4)</SelectItem>
                        <SelectItem value="VALE3">Vale (VALE3)</SelectItem>
                        <SelectItem value="CMIG4">Cemig (CMIG4)</SelectItem>
                        <SelectItem value="ABEV3">Ambev (ABEV3)</SelectItem>
                        <SelectItem value="BEEF3">Minerva (BEEF3)</SelectItem>
                        <SelectItem value="MGLU3">
                          Magazine Luiza (MGLU3)
                        </SelectItem>
                        <SelectItem value="PETZ3">
                          Pet Center (PETZ3)
                        </SelectItem>
                        <SelectItem value="RADL3">
                          Raia Drogasil (RADL3)
                        </SelectItem>
                        <SelectItem value="WEGE3">WEG (WEGE3)</SelectItem>
                        <SelectItem value="SUZB3">Suzano (SUZB3)</SelectItem>
                        <SelectItem value="RENT3">Localiza (RENT3)</SelectItem>
                        <SelectItem value="ELET3">
                          Eletrobras (ELET3)
                        </SelectItem>
                        <SelectItem value="GGBR4">Gerdau (GGBR4)</SelectItem>
                        <SelectItem value="CSNA3">CSN (CSNA3)</SelectItem>
                        <SelectItem value="KLBN11">Klabin (KLBN11)​</SelectItem>
                      </SelectGroup>
                    </SelectContent>
                  </Select>
                </FormControl>
                <FormDescription>As 20 ações mais negociadas.</FormDescription>
                <FormMessage />
              </FormItem>
            )}
          />
          <FormField
            control={form.control}
            name="side"
            render={({ field }) => (
              <FormItem>
                <FormLabel>Operação</FormLabel>
                <FormControl>
                  <Select
                    onValueChange={field.onChange}
                    defaultValue={field.value}
                  >
                    <SelectTrigger className="w-full">
                      <SelectValue placeholder="Selecione uma operação" />
                    </SelectTrigger>
                    <SelectContent>
                      <SelectGroup>
                        <SelectItem value="Compra">Compra</SelectItem>
                        <SelectItem value="Venda">Venda</SelectItem>
                      </SelectGroup>
                    </SelectContent>
                  </Select>
                </FormControl>
                <FormDescription>Operação de compra/venda.</FormDescription>
                <FormMessage />
              </FormItem>
            )}
          />
          <FormField
            control={form.control}
            name="price"
            render={({ field }) => (
              <FormItem>
                <FormLabel>Preço</FormLabel>
                <FormControl>
                  <Input
                    {...field}
                    type="text"
                    inputMode="numeric"
                    pattern="^(\s?R\$)?\s?\d{1,3}(\.\d{3})*(,\d{2})?$"
                    placeholder="R$ 10,99"
                    onChange={(e) =>
                      field.onChange(formatToBRL(e.target.value))
                    }
                    value={field.value}
                  />
                </FormControl>
                <FormDescription>Preço de negociação.</FormDescription>
                <FormMessage />
              </FormItem>
            )}
          />
          <FormField
            control={form.control}
            name="amount"
            render={({ field }) => (
              <FormItem>
                <FormLabel>Quantidade</FormLabel>
                <FormControl>
                  <Input
                    {...field}
                    value={field.value}
                    onChange={(e) =>
                      field.onChange(formatInteger(e.target.value))
                    }
                    type="text"
                    inputMode="numeric"
                    placeholder="2 un."
                  />
                </FormControl>
                <FormDescription>Digite um número de 1 a 1000.</FormDescription>
                <FormMessage />
              </FormItem>
            )}
          />
          <Button type="submit" disabled={loading}>
            {loading ? 'enviando' : 'Criar ordem'}
          </Button>
        </form>
      </Form>
    </div>
  );
}
