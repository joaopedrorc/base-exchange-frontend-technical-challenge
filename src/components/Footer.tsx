export function Footer() {
  return (
    <footer className="mt-auto w-full bg-slate-800 px-4 py-3 text-white">
      <div className="mx-auto flex max-w-7xl flex-col items-center justify-center text-sm sm:flex-row">
        <span>&copy; {new Date().getFullYear()} Gerenciador de Ordens</span>
      </div>
    </footer>
  );
}
