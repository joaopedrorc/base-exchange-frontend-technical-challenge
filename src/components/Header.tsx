import { Button } from '@/components/ui/button';
import { cn } from '@/lib/utils';

export function Header() {
  return (
    <header className="w-full bg-slate-800 px-4 py-3 text-white shadow-md">
      <div className="mx-auto flex max-w-7xl items-center justify-between">
        {/* Left: App Title */}
        <h1 className="text-xl font-semibold">Gerenciador de Ordens</h1>

        {/* Right: Buttons */}
        <div className="flex gap-2">
          <Button
            variant="secondary"
            className={cn('bg-slate-700 text-white hover:bg-slate-600')}
          >
            Home
          </Button>
          <Button className="bg-white text-slate-800 hover:bg-slate-100">
            Nova ordem
          </Button>
        </div>
      </div>
    </header>
  );
}
