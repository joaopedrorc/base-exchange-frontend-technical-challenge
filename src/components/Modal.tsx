'use client';

import { Button } from '@/components/ui/button';
import {
  Dialog,
  DialogContent,
  DialogDescription,
  DialogFooter,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
} from '@/components/ui/dialog';
import { useState } from 'react';
import { toast } from 'sonner';

export function Modal({
  orderId,
  orderStatus,
}: {
  orderId: string;
  orderStatus: string;
}) {
  const [open, setOpen] = useState<boolean>(false);

  const deleteSelectedOrder = async (orderId: string, orderStatus: string) => {
    if (orderStatus === 'Cancelada' || orderStatus === 'Executada') {
      setOpen(false);
      return toast.error(
        'Atenção: somente operações Abertas ou Parciais podem ser canceladas.'
      );
    }

    const response = await fetch('/api/orders', {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ id: orderId, status: 'Cancelada' }),
    });

    const data = await response.json();

    if (!response.ok) {
      toast.error(
        `Erro ao cancelar ordem: ${data.error}` || 'Erro ao cancelar ordem.'
      );
      setOpen(false);
      return;
    }

    toast.success('Ordem cancelada com sucesso!');

    setOpen(false);
  };

  return (
    <Dialog open={open} onOpenChange={setOpen}>
      <DialogTrigger asChild>
        <Button variant="destructive">Cancelar ordem</Button>
      </DialogTrigger>
      <DialogContent className="sm:max-w-[425px]">
        <DialogHeader>
          <DialogTitle>Cancelar ordem</DialogTitle>
          <DialogDescription>
            Tem certeza que deseja cancelar essa ordem?
          </DialogDescription>
        </DialogHeader>
        <DialogFooter>
          <Button variant="link" onClick={() => setOpen(false)}>
            Fechar
          </Button>
          <Button
            variant="destructive"
            onClick={() => deleteSelectedOrder(orderId, orderStatus)}
          >
            Cancelar
          </Button>
        </DialogFooter>
      </DialogContent>
    </Dialog>
  );
}
