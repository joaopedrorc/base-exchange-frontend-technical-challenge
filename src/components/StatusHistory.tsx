'use client';

import { useState } from 'react';
import { Button } from './ui/button';

interface StatusHistoryProps {
  id: string;
}

export function StatusHistory({ id }: StatusHistoryProps) {
  const [show, setShow] = useState(false);
  const [history, setHistory] = useState<string[] | null>(null);
  const [loading, setLoading] = useState(false);

  const fetchHistory = async () => {
    try {
      setLoading(true);
      const res = await fetch(`/api/status-history?id=${id}`);
      const data = await res.json();

      setHistory(data.history);
      setLoading(false);
    } catch (error) {
      console.log('ERROR', error);

      setLoading(false);
      setShow(false);
    }
  };

  const handleClick = () => {
    if (!show && !history) {
      fetchHistory();
    }
    setShow(!show);
  };

  return (
    <div>
      <Button className="bg-slate-800" onClick={handleClick}>
        {show ? 'Ocultar histórico' : 'Mostrar histórico de status'}
      </Button>

      {show && (
        <div className="mt-2 space-y-1 text-sm text-slate-700">
          {loading && <p>Carregando...</p>}
          {!loading && history?.length === 0 && (
            <p>Nenhum histórico encontrado.</p>
          )}
          {!loading &&
            history?.map((item, idx) => (
              <p key={idx} className="border-l-2 border-slate-300 pl-2">
                • {item}
              </p>
            ))}
        </div>
      )}
    </div>
  );
}
