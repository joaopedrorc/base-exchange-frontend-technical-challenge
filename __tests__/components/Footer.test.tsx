import { render } from '@testing-library/react';
import { Footer } from '@/components/Footer';

describe('Footer component', () => {
  it('should render Footer component', () => {
    render(<Footer />);
  });

  it('renders Footer component unchanged', async () => {
    const component = render(<Footer />);

    expect(component).toMatchSnapshot();
  });
});
