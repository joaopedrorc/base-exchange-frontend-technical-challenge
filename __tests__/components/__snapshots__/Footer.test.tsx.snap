// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer component renders Footer component unchanged 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <footer
        class="mt-auto w-full bg-slate-800 px-4 py-3 text-white"
      >
        <div
          class="mx-auto flex max-w-7xl flex-col items-center justify-center text-sm sm:flex-row"
        >
          <span>
            © 
            2025
             Gerenciador de Ordens
          </span>
        </div>
      </footer>
    </div>
  </body>,
  "container": <div>
    <footer
      class="mt-auto w-full bg-slate-800 px-4 py-3 text-white"
    >
      <div
        class="mx-auto flex max-w-7xl flex-col items-center justify-center text-sm sm:flex-row"
      >
        <span>
          © 
          2025
           Gerenciador de Ordens
        </span>
      </div>
    </footer>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
